service/test and pkg/proc tests are racy

What version of Delve are you using (dlv version)? 1.0.0-rc.1, git version
What version of Go are you using? (go version)? go1.8
What operating system and processor architecture are you using? linux/amd64

What did you do?
  go test -race github.com/derekparker/delve/pkg/proc
  go test -race github.com/derekparker/delve/service/test

What did you expect to see?
  test passes cleanly

What did you see instead?
  data races

Example data race messages: 

<pre>
==================
WARNING: DATA RACE
Read at 0x00c420292129 by goroutine 33:
  github.com/derekparker/delve/pkg/proc/native.(*Process).Running()
      /usr/local/google/home/hakim/go/src/github.com/derekparker/delve/pkg/proc/native/proc.go:130 +0xf0
  github.com/derekparker/delve/pkg/proc_test.TestHalt.func1.1()
      /usr/local/google/home/hakim/go/src/github.com/derekparker/delve/pkg/proc/proc_test.go:228 +0x52

Previous write at 0x00c420292129 by goroutine 25:
  github.com/derekparker/delve/pkg/proc/native.(*Thread).resumeWithSig()
      /usr/local/google/home/hakim/go/src/github.com/derekparker/delve/pkg/proc/native/threads_linux.go:43 +0x7a
  github.com/derekparker/delve/pkg/proc/native.(*Thread).resume()
      /usr/local/google/home/hakim/go/src/github.com/derekparker/delve/pkg/proc/native/threads_linux.go:39 +0x41
  github.com/derekparker/delve/pkg/proc/native.(*Process).resume()
      /usr/local/google/home/hakim/go/src/github.com/derekparker/delve/pkg/proc/native/proc_linux.go:386 +0x1f3
  github.com/derekparker/delve/pkg/proc/native.(*Process).ContinueOnce()
      /usr/local/google/home/hakim/go/src/github.com/derekparker/delve/pkg/proc/native/proc.go:266 +0x77
  github.com/derekparker/delve/pkg/proc.Continue()
      /usr/local/google/home/hakim/go/src/github.com/derekparker/delve/pkg/proc/proc.go:101 +0x5c
  github.com/derekparker/delve/pkg/proc_test.TestHalt.func1()
      /usr/local/google/home/hakim/go/src/github.com/derekparker/delve/pkg/proc/proc_test.go:237 +0x1eb
  github.com/derekparker/delve/pkg/proc_test.withTestProcess()
      /usr/local/google/home/hakim/go/src/github.com/derekparker/delve/pkg/proc/proc_test.go:78 +0x41a
  github.com/derekparker/delve/pkg/proc_test.withTestProcess()
      /usr/local/google/home/hakim/go/src/github.com/derekparker/delve/pkg/proc/proc_test.go:55 +0x572
  github.com/derekparker/delve/pkg/proc_test.TestHalt()
      /usr/local/google/home/hakim/go/src/github.com/derekparker/delve/pkg/proc/proc_test.go:253 +0xef
  testing.tRunner()
      /usr/lib/google-golang/src/testing/testing.go:657 +0x107

Goroutine 33 (running) created at:
  github.com/derekparker/delve/pkg/proc_test.TestHalt.func1()
      /usr/local/google/home/hakim/go/src/github.com/derekparker/delve/pkg/proc/proc_test.go:236 +0x1cd
  github.com/derekparker/delve/pkg/proc_test.withTestProcess()
      /usr/local/google/home/hakim/go/src/github.com/derekparker/delve/pkg/proc/proc_test.go:78 +0x41a
  github.com/derekparker/delve/pkg/proc_test.withTestProcess()
      /usr/local/google/home/hakim/go/src/github.com/derekparker/delve/pkg/proc/proc_test.go:55 +0x572
  github.com/derekparker/delve/pkg/proc_test.TestHalt()
      /usr/local/google/home/hakim/go/src/github.com/derekparker/delve/pkg/proc/proc_test.go:253 +0xef
  testing.tRunner()
      /usr/lib/google-golang/src/testing/testing.go:657 +0x107

Goroutine 25 (running) created at:
  testing.(*T).Run()
      /usr/lib/google-golang/src/testing/testing.go:697 +0x543
  testing.runTests.func1()
      /usr/lib/google-golang/src/testing/testing.go:884 +0xaa
  testing.tRunner()
      /usr/lib/google-golang/src/testing/testing.go:657 +0x107
  testing.runTests()
      /usr/lib/google-golang/src/testing/testing.go:890 +0x4e0
  testing.(*M).Run()
      /usr/lib/google-golang/src/testing/testing.go:824 +0x1c3
  github.com/derekparker/delve/pkg/proc/test.RunTestsWithFixtures()
      /usr/local/google/home/hakim/go/src/github.com/derekparker/delve/pkg/proc/test/support.go:89 +0x50
  github.com/derekparker/delve/pkg/proc_test.TestMain()
      /usr/local/google/home/hakim/go/src/github.com/derekparker/delve/pkg/proc/proc_test.go:45 +0xac
  main.main()
      github.com/derekparker/delve/pkg/proc/_test/_testmain.go:230 +0x20f
==================
==================
WARNING: DATA RACE
Read at 0x00c4206682a1 by goroutine 25:
  github.com/derekparker/delve/pkg/proc/native.(*Process).trapWait()
      /usr/local/google/home/hakim/go/src/github.com/derekparker/delve/pkg/proc/native/proc_linux.go:239 +0x4f2
  github.com/derekparker/delve/pkg/proc/native.(*Process).ContinueOnce()
      /usr/local/google/home/hakim/go/src/github.com/derekparker/delve/pkg/proc/native/proc.go:275 +0x25e
  github.com/derekparker/delve/pkg/proc.Continue()
      /usr/local/google/home/hakim/go/src/github.com/derekparker/delve/pkg/proc/proc.go:101 +0x5c
  github.com/derekparker/delve/pkg/proc_test.TestHalt.func1()
      /usr/local/google/home/hakim/go/src/github.com/derekparker/delve/pkg/proc/proc_test.go:237 +0x1eb
  github.com/derekparker/delve/pkg/proc_test.withTestProcess()
      /usr/local/google/home/hakim/go/src/github.com/derekparker/delve/pkg/proc/proc_test.go:78 +0x41a
  github.com/derekparker/delve/pkg/proc_test.withTestProcess()
      /usr/local/google/home/hakim/go/src/github.com/derekparker/delve/pkg/proc/proc_test.go:55 +0x572
  github.com/derekparker/delve/pkg/proc_test.TestHalt()
      /usr/local/google/home/hakim/go/src/github.com/derekparker/delve/pkg/proc/proc_test.go:253 +0xef
  testing.tRunner()
      /usr/lib/google-golang/src/testing/testing.go:657 +0x107

Previous write at 0x00c4206682a1 by goroutine 33:
  github.com/derekparker/delve/pkg/proc/native.(*Process).RequestManualStop()
      /usr/local/google/home/hakim/go/src/github.com/derekparker/delve/pkg/proc/native/proc.go:190 +0xcb
  github.com/derekparker/delve/pkg/proc_test.TestHalt.func1.1()
      /usr/local/google/home/hakim/go/src/github.com/derekparker/delve/pkg/proc/proc_test.go:229 +0x74

Goroutine 25 (running) created at:
  testing.(*T).Run()
      /usr/lib/google-golang/src/testing/testing.go:697 +0x543
  testing.runTests.func1()
      /usr/lib/google-golang/src/testing/testing.go:884 +0xaa
  testing.tRunner()
      /usr/lib/google-golang/src/testing/testing.go:657 +0x107
  testing.runTests()
      /usr/lib/google-golang/src/testing/testing.go:890 +0x4e0
  testing.(*M).Run()
      /usr/lib/google-golang/src/testing/testing.go:824 +0x1c3
  github.com/derekparker/delve/pkg/proc/test.RunTestsWithFixtures()
      /usr/local/google/home/hakim/go/src/github.com/derekparker/delve/pkg/proc/test/support.go:89 +0x50
  github.com/derekparker/delve/pkg/proc_test.TestMain()
      /usr/local/google/home/hakim/go/src/github.com/derekparker/delve/pkg/proc/proc_test.go:45 +0xac
  main.main()
      github.com/derekparker/delve/pkg/proc/_test/_testmain.go:230 +0x20f

Goroutine 33 (running) created at:
  github.com/derekparker/delve/pkg/proc_test.TestHalt.func1()
      /usr/local/google/home/hakim/go/src/github.com/derekparker/delve/pkg/proc/proc_test.go:236 +0x1cd
  github.com/derekparker/delve/pkg/proc_test.withTestProcess()
      /usr/local/google/home/hakim/go/src/github.com/derekparker/delve/pkg/proc/proc_test.go:78 +0x41a
  github.com/derekparker/delve/pkg/proc_test.withTestProcess()
      /usr/local/google/home/hakim/go/src/github.com/derekparker/delve/pkg/proc/proc_test.go:55 +0x572
  github.com/derekparker/delve/pkg/proc_test.TestHalt()
      /usr/local/google/home/hakim/go/src/github.com/derekparker/delve/pkg/proc/proc_test.go:253 +0xef
  testing.tRunner()
      /usr/lib/google-golang/src/testing/testing.go:657 +0x107
==================
--- FAIL: TestHalt (1.03s)
        testing.go:610: race detected during execution of test
</pre>

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -5,6 +5,12 @@ This project adheres to Semantic Versioning.
 
 All changes mention the author, unless contributed by me (@derekparker).
 
+## [RELEASE TO BE DEFINED] DATE TO BE DEFINED
+
+### Fixed
+
+- Data races in tests (@aarzilli)
+
 ## [1.0.0-rc.2] DATE TO BE DEFINED
 
 ### Added

--- a/pkg/proc/core/core.go
+++ b/pkg/proc/core/core.go
@@ -304,8 +304,7 @@ func (p *Process) Pid() int {
 	return p.core.Pid
 }
 
-func (p *Process) Running() bool {
-	return false
+func (p *Process) ResumeNotify(chan<- struct{}) {
 }
 
 func (p *Process) SelectedGoroutine() *proc.G {

--- a/pkg/proc/gdbserial/gdbserver.go
+++ b/pkg/proc/gdbserial/gdbserver.go
@@ -490,8 +490,8 @@ func (p *Process) Exited() bool {
 	return p.exited
 }
 
-func (p *Process) Running() bool {
-	return p.conn.running
+func (p *Process) ResumeNotify(ch chan<- struct{}) {
+	p.conn.resumeChan = ch
 }
 
 func (p *Process) FindThread(threadID int) (proc.Thread, bool) {

--- a/pkg/proc/gdbserial/gdbserver_conn.go
+++ b/pkg/proc/gdbserial/gdbserver_conn.go
@@ -23,7 +23,8 @@ type gdbConn struct {
 	inbuf  []byte
 	outbuf bytes.Buffer
 
-	running bool
+	running    bool
+	resumeChan chan<- struct{}
 
 	direction proc.Direction // direction of execution
 
@@ -543,6 +544,10 @@ func (conn *gdbConn) resume(sig uint8, tu *threadUpdater) (string, uint8, error)
 	defer func() {
 		conn.running = false
 	}()
+	if conn.resumeChan != nil {
+		close(conn.resumeChan)
+		conn.resumeChan = nil
+	}
 	return conn.waitForvContStop("resume", "-1", tu)
 }
 

--- a/pkg/proc/interface.go
+++ b/pkg/proc/interface.go
@@ -6,6 +6,11 @@ import (
 
 // Process represents the target of the debugger. This
 // target could be a system process, core file, etc.
+//
+// Implementations of Process are not required to be thread safe and users
+// of Process should not assume they are.
+// There is one exception to this rule: it is safe to call RequestManualStop
+// concurrently with ContinueOnce.
 type Process interface {
 	Info
 	ProcessManipulation

--- a/pkg/proc/interface.go
+++ b/pkg/proc/interface.go
@@ -52,8 +52,10 @@ type Checkpoint struct {
 // Info is an interface that provides general information on the target.
 type Info interface {
 	Pid() int
+	// ResumeNotify specifies a channel that will be closed the next time
+	// ContinueOnce finishes resuming the target.
+	ResumeNotify(chan<- struct{})
 	Exited() bool
-	Running() bool
 	BinInfo() *BinaryInfo
 
 	ThreadInfo

--- a/pkg/proc/native/proc.go
+++ b/pkg/proc/native/proc.go
@@ -37,6 +37,7 @@ type Process struct {
 	breakpointIDCounter         int
 	internalBreakpointIDCounter int
 	firstStart                  bool
+	haltMu                      sync.Mutex
 	halt                        bool
 	exited                      bool
 	ptraceChan                  chan func()
@@ -187,6 +188,8 @@ func (dbp *Process) RequestManualStop() error {
 	if dbp.exited {
 		return &proc.ProcessExitedError{}
 	}
+	dbp.haltMu.Lock()
+	defer dbp.haltMu.Unlock()
 	dbp.halt = true
 	return dbp.requestManualStop()
 }

--- a/pkg/proc/native/proc.go
+++ b/pkg/proc/native/proc.go
@@ -39,6 +39,7 @@ type Process struct {
 	firstStart                  bool
 	haltMu                      sync.Mutex
 	halt                        bool
+	resumeChan                  chan<- struct{}
 	exited                      bool
 	ptraceChan                  chan func()
 	ptraceDoneChan              chan interface{}
@@ -89,11 +90,6 @@ func (dbp *Process) Detach(kill bool) (err error) {
 		dbp.bi.Close()
 		return nil
 	}
-	if dbp.Running() {
-		if err = dbp.Halt(); err != nil {
-			return
-		}
-	}
 	if !kill {
 		// Clean up any breakpoints we've set.
 		for _, bp := range dbp.breakpoints {
@@ -124,15 +120,8 @@ func (dbp *Process) Exited() bool {
 	return dbp.exited
 }
 
-// Running returns whether the debugged
-// process is currently executing.
-func (dbp *Process) Running() bool {
-	for _, th := range dbp.threads {
-		if th.running {
-			return true
-		}
-	}
-	return false
+func (dbp *Process) ResumeNotify(ch chan<- struct{}) {
+	dbp.resumeChan = ch
 }
 
 func (dbp *Process) Pid() int {
@@ -273,6 +262,11 @@ func (dbp *Process) ContinueOnce() (proc.Thread, error) {
 	dbp.allGCache = nil
 	for _, th := range dbp.threads {
 		th.clearBreakpointState()
+	}
+
+	if dbp.resumeChan != nil {
+		close(dbp.resumeChan)
+		dbp.resumeChan = nil
 	}
 
 	trapthread, err := dbp.trapWait(-1)

--- a/pkg/proc/native/proc_darwin.go
+++ b/pkg/proc/native/proc_darwin.go
@@ -291,7 +291,10 @@ func (dbp *Process) trapWait(pid int) (*Thread, error) {
 			return nil, proc.ProcessExitedError{Pid: dbp.pid, Status: status.ExitStatus()}
 
 		case C.MACH_RCV_INTERRUPTED:
-			if !dbp.halt {
+			dbp.haltMu.Lock()
+			halt := dbp.halt
+			dbp.haltMu.Unlock()
+			if !halt {
 				// Call trapWait again, it seems
 				// MACH_RCV_INTERRUPTED is emitted before
 				// process natural death _sometimes_.
@@ -319,7 +322,10 @@ func (dbp *Process) trapWait(pid int) (*Thread, error) {
 		dbp.updateThreadList()
 		th, ok := dbp.threads[int(port)]
 		if !ok {
-			if dbp.halt {
+			dbp.haltMu.Lock()
+			halt := dbp.halt
+			dbp.haltMu.Unlock()
+			if halt {
 				dbp.halt = false
 				return th, nil
 			}

--- a/pkg/proc/native/proc_linux.go
+++ b/pkg/proc/native/proc_linux.go
@@ -236,7 +236,10 @@ func (dbp *Process) trapWait(pid int) (*Thread, error) {
 			// Sometimes we get an unknown thread, ignore it?
 			continue
 		}
-		if status.StopSignal() == sys.SIGTRAP && dbp.halt {
+		dbp.haltMu.Lock()
+		halt := dbp.halt
+		dbp.haltMu.Unlock()
+		if status.StopSignal() == sys.SIGTRAP && halt {
 			th.running = false
 			dbp.halt = false
 			return th, nil

--- a/pkg/proc/proc_test.go
+++ b/pkg/proc/proc_test.go
@@ -213,27 +213,22 @@ func TestHalt(t *testing.T) {
 		_, err := setFunctionBreakpoint(p, "main.loop")
 		assertNoError(err, t, "SetBreakpoint")
 		assertNoError(proc.Continue(p), t, "Continue")
-		if p.Running() {
-			t.Fatal("process still running")
-		}
 		if p, ok := p.(*native.Process); ok {
 			for _, th := range p.ThreadList() {
 				_, err := th.Registers(false)
 				assertNoError(err, t, "Registers")
 			}
 		}
+		resumeChan := make(chan struct{})
 		go func() {
-			for {
-				time.Sleep(100 * time.Millisecond)
-				if p.Running() {
-					if err := p.RequestManualStop(); err != nil {
-						t.Fatal(err)
-					}
-					stopChan <- nil
-					return
-				}
+			<-resumeChan
+			time.Sleep(100 * time.Millisecond)
+			if err := p.RequestManualStop(); err != nil {
+				t.Fatal(err)
 			}
+			stopChan <- nil
 		}()
+		p.ResumeNotify(resumeChan)
 		assertNoError(proc.Continue(p), t, "Continue")
 		<-stopChan
 		// Loop through threads and make sure they are all
@@ -601,9 +596,6 @@ func TestNextNetHTTP(t *testing.T) {
 	}
 	withTestProcess("testnextnethttp", t, func(p proc.Process, fixture protest.Fixture) {
 		go func() {
-			for !p.Running() {
-				time.Sleep(50 * time.Millisecond)
-			}
 			// Wait for program to start listening.
 			for {
 				conn, err := net.Dial("tcp", "localhost:9191")
@@ -1934,10 +1926,6 @@ func TestIssue462(t *testing.T) {
 	}
 	withTestProcess("testnextnethttp", t, func(p proc.Process, fixture protest.Fixture) {
 		go func() {
-			for !p.Running() {
-				time.Sleep(50 * time.Millisecond)
-			}
-
 			// Wait for program to start listening.
 			for {
 				conn, err := net.Dial("tcp", "localhost:9191")

--- a/pkg/proc/proc_unix_test.go
+++ b/pkg/proc/proc_unix_test.go
@@ -25,24 +25,21 @@ func TestIssue419(t *testing.T) {
 		_, err := setFunctionBreakpoint(p, "main.main")
 		assertNoError(err, t, "SetBreakpoint()")
 		assertNoError(proc.Continue(p), t, "Continue()")
+		resumeChan := make(chan struct{})
 		go func() {
-			for {
-				time.Sleep(500 * time.Millisecond)
-				if p.Running() {
-					time.Sleep(2 * time.Second)
-					if p.Pid() <= 0 {
-						// if we don't stop the inferior the test will never finish
-						p.RequestManualStop()
-						p.Kill()
-						t.Fatalf("Pid is zero or negative: %d", p.Pid())
-						return
-					}
-					err := syscall.Kill(p.Pid(), syscall.SIGINT)
-					assertNoError(err, t, "syscall.Kill")
-					return
-				}
+			time.Sleep(500 * time.Millisecond)
+			<-resumeChan
+			if p.Pid() <= 0 {
+				// if we don't stop the inferior the test will never finish
+				p.RequestManualStop()
+				p.Kill()
+				t.Fatalf("Pid is zero or negative: %d", p.Pid())
+				return
 			}
+			err := syscall.Kill(p.Pid(), syscall.SIGINT)
+			assertNoError(err, t, "syscall.Kill")
 		}()
+		p.ResumeNotify(resumeChan)
 		err = proc.Continue(p)
 		if _, exited := err.(proc.ProcessExitedError); !exited {
 			t.Fatalf("Unexpected error after Continue(): %v\n", err)

--- a/service/debugger/debugger.go
+++ b/service/debugger/debugger.go
@@ -200,9 +200,6 @@ func (d *Debugger) Restart(pos string) ([]api.DiscardedBreakpoint, error) {
 	}
 
 	if !d.target.Exited() {
-		if d.target.Running() {
-			d.target.Halt()
-		}
 		// Ensure the process is in a PTRACE_STOP.
 		if err := stopProcess(d.ProcessPid()); err != nil {
 			return nil, err

